sys/can/dll.c

```diff
@@ -0,0 +1,460 @@
+
+    msg_send_receive(&msg, &reply, candev_list[ifnum]);
+
+    can_pkt_free(pkt);
+
+    return 0;
+}
+
+static int register_filter_entry(can_reg_entry_t *entry, struct can_filter *filter, void *param)
+{
+    msg_t msg, reply;
+    int ret;
+
+    DEBUG("register_filter_entry: ifnum=%d, filter=0x%" PRIx32 ", mask=0x%" PRIx32 ", param=%p\n",
+          entry->ifnum, filter->can_id, filter->can_mask, param);
+
+    ret = can_router_register(entry, filter->can_id, filter->can_mask, param);
+    if (ret < 0) {
+        return -ENOMEM;
+    }
+    else if (ret == 1) {
+        DEBUG("raw_can_subscribe_rx: filter=0x%" PRIx32 " already in use\n", filter->can_id);
+        return 0;
+    }
+
+    msg.type = CAN_MSG_SET_FILTER;
+    msg.content.ptr = filter;
+    msg_send_receive(&msg, &reply, candev_list[entry->ifnum]);
+
+    if ((int) reply.content.value < 0) {
+        can_router_unregister(entry, filter->can_id, filter->can_mask, param);
+        return -ENOMEM;
+    }
+
+    return 0;
+}
+
+static int unregister_filter_entry(can_reg_entry_t *entry, struct can_filter *filter, void *param)
+{
+    msg_t msg, reply;
+    int ret;
+
+    DEBUG("unregister_filter_entry: ifnum=%d, filter=0x%" PRIx32 ", mask=0x%" PRIx32 ", param=%p\n",
+          entry->ifnum, filter->can_id, filter->can_mask, param);
+
+    ret = can_router_unregister(entry, filter->can_id, filter->can_mask, param);
+    if (ret < 0) {
+        return -ENOMEM;
+    }
+    else if (ret == 1) {
+        DEBUG("raw_can_unsubscribe_rx: filter=0x%" PRIx32 " still in use\n", filter->can_id);
+        return 0;
+    }
+
+    msg.type = CAN_MSG_REMOVE_FILTER;
+    msg.content.ptr = filter;
+    msg_send_receive(&msg, &reply, candev_list[entry->ifnum]);
+
+    if ((int) reply.content.value < 0) {
+        return -ENOMEM;
+    }
+
+    return 0;
+}
+
+int raw_can_subscribe_rx(unsigned int ifnum, struct can_filter *filter, kernel_pid_t pid, void *param)
+{
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    if (filter == NULL) {
+        return -ENOENT;
+    }
+
+    can_reg_entry_t entry;
+    entry.ifnum = ifnum;
+    entry.target.pid = pid;
+#ifdef MODULE_CAN_MBOX
+    entry.type = CAN_TYPE_DEFAULT;
+#endif
+
+    return register_filter_entry(&entry, filter, param);
+}
+
+#ifdef MODULE_CAN_MBOX
+int raw_can_subscribe_rx_mbox(unsigned int ifnum, struct can_filter *filter, mbox_t *mbox, void *param)
+{
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    if (filter == NULL) {
+        return -ENOENT;
+    }
+
+    can_reg_entry_t entry;
+    entry.ifnum = ifnum;
+    entry.target.mbox = mbox;
+    entry.type = CAN_TYPE_MBOX;
+
+    return register_filter_entry(&entry, filter, param);
+}
+#endif
+
+int raw_can_unsubscribe_rx(unsigned int ifnum, struct can_filter *filter, kernel_pid_t pid, void *param)
+{
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    if (filter == NULL) {
+        return -ENOENT;
+    }
```

Comment:
I don't have any use case in mind where it could be really usefull to distinguish error codes, except for debugging. I'm also wondering if I should not replace all theses tests by simple asserts.